[docker-ptf]: TACACS fix, Python 3.9 deprecation fixes and build updates

opcoder0 opened the pull request:
#### Why I did it

With tacacs+ built manually for PY3 only image the directory `/etc/tacacs+` and `/etc/default/tacacs+` are missing causing tests to fail.
The most recent changes to pip packages throws errors related to Python 3.9 deprecations like -
- `from distutils.command.build_py import build_py_2to3 as _build_py` - Python 3.9 does not have `build_py_2to3` anymore.
- `AttributeError: 'HTMLParser' object has no attribute 'unescape'` when installing packages like nnpy, ipython etc. - Fixed by using virtualenv instead.

These errors are because some packages downgrade setuptools causing subsequent package installs to fail.

##### Work item tracking
- Microsoft ADO **(number only)**: 29869339

#### How I did it

Update Dockerfile for docker-ptf only where it is applicable to "py3".

#### How to verify it

Build and ran tests (verifications still in progress)

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

Not applicable. The image is not built on any branch yet. The flag is turned off by default.

#### Description for the changelog

[docker-ptf]: TACACS fix, Python 3.9 deprecation fixes and build updates

With tacacs+ built manually for PY3 only image the directory `/etc/tacacs+` and `/etc/default/tacacs+` are missing causing tests to fail.
The most recent changes to pip packages throws errors related to Python 3.9 deprecations like -
- `from distutils.command.build_py import build_py_2to3 as _build_py` - Python 3.9 does not have `build_py_2to3` anymore.
- `AttributeError: 'HTMLParser' object has no attribute 'unescape'` when installing packages like nnpy, ipython etc. - Fixed by using virtualenv instead.

These errors are because some packages downgrade setuptools causing subsequent package installs to fail.

#### Link to config_db schema for YANG module changes

Not applicable

#### A picture of a cute animal (not mandatory but encouraged)

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

wangxin commented on dockers/docker-ptf/Dockerfile.j2:
```diff
@@ -192,23 +196,40 @@ ENV LANG=C.UTF-8 LC_ALL=C.UTF-8 PYTHONIOENCODING=UTF-8
     && pip3 install ipython==5.4.1 \
+{% if PTF_ENV_PY_VER == "py3" %}
+    && pip3 install setuptools==51.0.0 \
+{% endif %}
```
Why not install all of the offending packages and lastly install setuptools once?
It looks tedious to install one offending package, install setuptools, and then repeat for another offending package.

opcoder0 replied:
Each offending package installed leaves `setuptools` in a version the other offending package does not like, and it fails with an obscure unrelated looking error like this. This is the install of `ipython` after `supervisor` which had downgraded setuptools to 40.8 -

```
Collecting ipython==5.4.1                                                                                                                                                                                                                                       
  Using cached ipython-5.4.1-py3-none-any.whl.metadata (4.3 kB)
...
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [12 lines of output]
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 14, in <module>
        File "/usr/local/lib/python3.9/dist-packages/setuptools/__init__.py", line 20, in <module>
          from setuptools.dist import Distribution, Feature
        File "/usr/local/lib/python3.9/dist-packages/setuptools/dist.py", line 35, in <module>
          from setuptools.depends import Require
        File "/usr/local/lib/python3.9/dist-packages/setuptools/depends.py", line 7, in <module>
          from .py33compat import Bytecode
        File "/usr/local/lib/python3.9/dist-packages/setuptools/py33compat.py", line 55, in <module>
          unescape = getattr(html, 'unescape', html_parser.HTMLParser().unescape)
      AttributeError: 'HTMLParser' object has no attribute 'unescape'
      [end of output]
```

wangxin commented on dockers/docker-ptf/Dockerfile.j2:
```diff
@@ -217,6 +238,9 @@ RUN pip3 install setuptools \
     && pip3 install grpcio-tools \
+{% if PTF_ENV_PY_VER == "py3" %}
+    && pip3 install setuptools==51.0.0 \
+{% endif %}
```
This pattern has been repeated multiple times. Is there a better way to install setuptools only once in the very end?

opcoder0 replied:
I have tried to resolve this by using a j2 macro instead.

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

opcoder0 commented:
@wangxin - based on your previous comment around many j2 if checks in the Dockerfile. I have refactored and cleaned up the code and used macros to make the code readable. As the diff is large compared to master I have created a PR to the file you commented on - https://github.com/opcoder0/sonic-buildimage/pull/1. The change will have to be tested against mixed and py3 as it changes both sides of the code.

liushilongbuaa commented:
@opcoder0 , do you have plan to upgrade docker-ptf to bookworm?
Currently it is built in buster env. Almost all dockers are upgraded to bookworm.

wangxin commented:
@opcoder0 Can you resolve the merge conflict?

opcoder0 commented:
> @opcoder0 Can you resolve the merge conflict?

Resolved.

mssonicbld commented:
/azp run Azure.sonic-buildimage

saiarcot895 commented on dockers/docker-ptf/Dockerfile.j2:
```diff
@@ -195,14 +199,33 @@ ENV LANG=C.UTF-8 LC_ALL=C.UTF-8 PYTHONIOENCODING=UTF-8
+        && pip3 install supervisor \
+        && pip3 install ipython==5.4.1 \
+        && pip3 install exabgp \
+        && pip3 install grpcio-tools
```
`scapy==2.5.0` and `thrift` are missing from this list.
